### ISSUE TITLE
Include fuzz test case logging

### ISSUE THREAD

**Issue by a-t-0** [TruCol/Solidity-Forge-Template#2]
- [ ] Remove unused code.
- [ ] Do a round of simplification and refactoring.
- [ ] Ask on how to simplify code further.
- [ ] Determine if you can directly log the variable names instead of abc etc. If not, determine if you can overwrite the text in the file after all runs, or if not, see if you can create a duplicate file with the same counts and with variable names instead of abc etc.
- [ ] If you can log variable names directly, implement it, and afterwards do another round of refactoring and simplification.
- [ ] If you can log variable names directly, after the refactoring and simplification, as advice on how to simplify further.
- [ ] See if you can include a hard assert on whether each test case is hit often enough.
- [ ] See if you can visualise the hitrates of the relevant test cases in jacoco with branch coverage depth or something.
- [ ] Improve the explanation.
- [ ] Ask if something like this already exists, or if not, if someone would like to use it in some mainstream repo/foundry.

**Comment by a-t-0** [comment]
A string replacement function is found and included with a small test.

Now, one can create an addtional file (at the ends of the fuzz test) that contains the actual hit rates with the human readable names.  Or one can apply the replacement function to switch back to `abc` mapping key notation when the file is read in (though that seems to be a lot of file read/write operations in the fuzz test. 

- [ ] Determine the exact directory structure in which you want to be logging, and ensure it also works if the tests are ran in parallel.
- [ ] Implement the logging structure.
- [ ] write a function that can copy a file content. Then ensure you can replace the strings in the copied file content.
- [ ] See if you can use the `afterAll` function for invariant tests.
- [ ] call the copy file and replace contents function afterAll.

**Comment by a-t-0** [comment]
- [ ] // TODO: find out the first empty place, and put it there.
- [ ] Determine why the pre-commit changes:
```sol
abi.encodePacked(
            '{ "hitCount":',
            Strings.toString(values[i].number),
            ', "variableName": "',
            values[i].str,
            '"}'
          )
```
to:
```diff
@@ -36,11 +36,11 @@ contract TestCaseHitRateLoggerToFile is PRBTest, StdCheats {
       for (uint256 i = 0; i < keys.length - 1; i++) {
         tupleStrings[i] = string(
           abi.encodePacked(
-            '{ "hitCount":',
+            "{ 'hitCount':",
             Strings.toString(values[i].number),
-            ', "variableName": "',
+            ", 'variableName': '",
             values[i].str,
-            '"}'
+            "'}"
           )
         );
 
```

**Comment by a-t-0** [comment]
Fixed exporting tuple to file. 
- [ ] Generalise it to exporting array of tuples to file using: apple array from: https://book.getfoundry.sh/cheatcodes/parse-json (and also look at: https://book.getfoundry.sh/cheatcodes/serialize-json) in a separate file ending with "C"
- [ ] Re-think if you can directly export the struct with the array of tuples from any test file. 
    - [ ] If yes: Write example that directly exports the array of tuples from afile.
    - [ ] If no: Ensure you can export import the map to from and to file.
- [ ] Move as much as you can into (new helper).

**Comment by a-t-0** [comment]
-  [ ] Set the required hit rates in the test file.
- [ ] Ensure the required hit rates are propagated to the initialisation of the map.
- [ ] Verify the increments are applied to the correct position in the map. Perhaps by setting the map key to the variable name, even though that would be double up. (Perhaps this is correct for the mapping, but not needed for the struct. Check how this relation is managed between the map and struct creation and reading.
- [ ] Once it is remove unneeded functions.

**Comment by a-t-0** [comment]
- [ ] Restore pre-commit compliance.
- [ ] Get the total number of runs from the `foundry.toml` file.
- [ ] Add an assert for equality of total nr of runs and counted runs (fix +4 over specified nr of runs).
- [ ] Facilitate multiple fuzz tests files to run simultaneously.
- [ ] Optionally merge all logging into a single file.
- [ ] Optionally allow multiple fuzz test functions per file, with logging.
- [ ] Add assert  `requiredHitCount` is reached for each variable.
- [ ] Optionally include overrideNrOfFuzzRuns in constants to prevent throwing error if manual override for debugging.

**Comment by a-t-0** [comment]
- [ ] Move assertCoverage into separate file and generalise it for any generic number of parameters.

**Comment by a-t-0** [comment]
In the invariant clone adoption the lib directory is different which yields compilation error.
- [ ]  Determine how to automatically generate the desired lib directory.
- [ ] Verify the auto-generated lib directory allows for compiling on the pasted src/ test and script dir.
- [ ] Update the readme with the new prerequisites required to generate the lib directory
- [ ] See if other old prerequisites can be deleted.
- [ ] Ensure the setup becomes pre-commit compliant.
- [ ]
